src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs

```diff
@@ -42,13 +43,28 @@ public static partial class JsonSerializer
                         Debug.Assert(state.Current.JsonClassInfo != default);
 
                         ReadOnlySpan<byte> propertyName = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;
-                        state.Current.JsonPropertyInfo = state.Current.JsonClassInfo.GetProperty(options, propertyName, ref state.Current);
-                        if (state.Current.JsonPropertyInfo == null)
+
+                        if (state.Current.IsDictionary())
                         {
-                            state.Current.JsonPropertyInfo = s_missingProperty;
+                            string keyName = reader.GetString();
+
+                            if (options.DictionaryKeyPolicy != null)
+                            {
+                                keyName = options.DictionaryKeyPolicy.ConvertName(keyName);
+                            }
+
+                            state.Current.TempDictKeys.Add(keyName);
```

Comment:
Per other comments, if we remove TempDictKeys, this would create a JsonPropertyInfo<TValue> - ideally we could use the same JsonPropertyInfo instance for all entries in the dictionary (for performance)